### PR TITLE
add_ci.tbl_svysummary()

### LINKED ISSUE #965: Add a `add_ci.tbl_svysummary()` function
Use `survey::svyciprop()` to calculate CIs

### PR DESCRIPTION
**Pull request by larmarange**
Extend add_ci() to deal with tbl_svysummary tables

fix #965 

- [x] add a way to customize the number of degrees of freedom
- [x] add examples
- [x] add unit test

--------------------------------------------------------------------------------

Reviewer Checklist (if item does not apply, mark is as complete)

- [x] Ensure all package dependencies are installed by running `renv::install()`
- [x] PR branch has pulled the most recent updates from master branch. Ensure the pull request branch and your local version match and both have the latest updates from the master branch.
- [x] If an update was made to `tbl_summary()`, was the same change implemented for `tbl_svysummary()`?
- [x] If a new function was added, function included in `_pkgdown.yml`
- [x] If a bug was fixed, a unit test was added for the bug check
- [x] Run `pkgdown::build_site()`. Check the R console for errors, and review the rendered website.
- [x] Code coverage is suitable for any new functions/features. Review coverage with `withr::with_envvar(new = c("NOT_CRAN" = "true"), covr::report())`. Begin in a fresh R session without any packages loaded. 
- [x] R CMD Check runs without errors, warnings, and notes
- [x] `usethis::use_spell_check()` runs with no spelling errors in documentation

When the branch is ready to be merged into master:
- [x] Update `NEWS.md` with the changes from this pull request under the heading "`# gtsummary (development version)`". If there is an issue associated with the pull request, reference it in parentheses at the end update (see `NEWS.md` for examples).
- [x] Increment the version number using `usethis::use_version(which = "dev")` 
- [x] Run `codemetar::write_codemeta()`
- [x] Run `usethis::use_spell_check()` again
- [x] Approve Pull Request
- [x] Merge the PR. Please use "Squash and merge".

### CODE REVIEW

**Comment by larmarange** [comment]
First draft. Support for mean, median and proportion CIs in `tbl_svysummary` tables.

``` r
library(gtsummary)
library(survey)
#> Le chargement a nécessité le package : grid
#> Le chargement a nécessité le package : Matrix
#> Le chargement a nécessité le package : survival
#> 
#> Attachement du package : 'survey'
#> L'objet suivant est masqué depuis 'package:graphics':
#> 
#>     dotchart

data(api, package = "survey")

tbl <-
  survey::svydesign(id = ~dnum, weights = ~pw, data = apiclus1, fpc = ~fpc) %>%
  tbl_svysummary(
    include = c(api00, stype),
    statistic = all_continuous() ~ "{mean}"
  )
tbl %>% 
  add_ci() %>%
  as_kable()
```

| **Characteristic** | **N = 6,194** | **95% CI** |
|:-------------------|:-------------:|:----------:|
| api00              |      644      |  598, 690  |
| stype              |               |            |
| E                  |  4,874 (79%)  |  67%, 87%  |
| H                  |  474 (7.7%)   | 3.5%, 16%  |
| M                  |   846 (14%)   | 8.4%, 21%  |

``` r


tbl <-
  survey::svydesign(id = ~dnum, weights = ~pw, data = apiclus1, fpc = ~fpc) %>%
  tbl_svysummary(
    include = c(api00, stype),
    by = both,
    statistic = all_continuous() ~ "{mean}"
  )
tbl %>% 
  add_overall() %>%
  add_ci() %>%
  as_kable()
```

| **Characteristic** | **Overall**, N = 6,194 | **95% CI** | **No**, N = 1,692 | **95% CI** | **Yes**, N = 4,502 | **95% CI** |
|:-------------------|:----------------------:|:----------:|:-----------------:|:----------:|:------------------:|:----------:|
| api00              |          644           |  598, 690  |        633        |  575, 691  |        648         |  604, 692  |
| stype              |                        |            |                   |            |                    |            |
| E                  |      4,874 (79%)       |  67%, 87%  |    1,083 (64%)    |  42%, 81%  |    3,791 (84%)     |  73%, 91%  |
| H                  |       474 (7.7%)       | 3.5%, 16%  |     237 (14%)     | 6.5%, 28%  |     237 (5.3%)     | 2.1%, 13%  |
| M                  |       846 (14%)        | 8.4%, 21%  |     372 (22%)     | 8.5%, 46%  |     474 (11%)      | 5.6%, 19%  |

``` r

tbl <-
  survey::svydesign(id = ~dnum, weights = ~pw, data = apiclus1, fpc = ~fpc) %>%
  tbl_svysummary(
    include = c(api00),
    by = both
  )
tbl %>% 
  add_ci(method = all_continuous() ~ "svymedian") %>%
  add_overall() %>%
  as_kable()
```

| **Characteristic** | **Overall**, N = 6,194 | **No**, N = 1,692 | **95% CI** | **Yes**, N = 4,502 | **95% CI** |
|:-------------------|:----------------------:|:-----------------:|:----------:|:------------------:|:----------:|
| api00              |     652 (552, 718)     |  631 (556, 710)   |  547, 722  |   654 (551, 722)   |  566, 717  |

<sup>Created on 2022-12-14 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by larmarange** [comment]
So far, the number of degrees of freedom cannot be customized.  It would require an additional argument in `add_ci()`, or we could use eventually `...`.

Note: `svyciprop()` and `svyquantile()` are using `degf(design)` while `confint(svymean())` use `Inf`.

**Comment by ddsjoberg** [comment]
awesome! thanks for the addition @larmarange !

**Comment by larmarange** [comment]
@ddsjoberg  I think that the PR is ready for review. Cheers

**Review by ddsjoberg — APPROVED** [review]
Great update that many users will be excited about!

**Comment by aspina7** [comment]
hi @larmarange - thanks so much for this pr, its fantastic! 

minor question as not sure if I am missing something - `add_ci()` throws an error with logical vars (also not dichotomous numeric) but functions if changed to a character. 
It's easy enough to just mutate to char before running so no issue really - just wondering if there is a methodological reason behind this? 

Thanks again! 


``` r
library(gtsummary)
#> Warning: Paket 'gtsummary' wurde unter R Version 4.2.2 erstellt

data(api, package = "survey")

## make a logical var 
apiclus1 <- mutate(apiclus1, awards2 = awards == "Yes")
head(apiclus1$awards2)
#> [1]  TRUE  TRUE  TRUE FALSE  TRUE  TRUE

## does not work with binary 
tbl <-
  survey::svydesign(id = ~dnum, weights = ~pw, data = apiclus1, fpc = ~fpc) %>%
  tbl_svysummary(
    include = c(awards2),
    statistic = all_continuous() ~ "{mean}"
  )

tbl %>% 
  add_ci() %>% 
  as_kable()
#> There was an error for variable 'awards2':
#> Error in `mutate()`:
#> ! Problem while computing `ci = calculate_svyprop_ci(...)`.
#> ℹ The error occurred in row 1.
#> Caused by error in `.data$variable_levels`:
#> ! Column `variable_levels` not found in `.data`.
#> ℹ No columns were selected. Use `quiet = TRUE` to supress these messages.
#> ℹ As a usage guide, the code below re-creates the current column headers.
#> modify_header(
#>   label = "**Characteristic**",
#>   stat_0 = "**N = 6,194**",
#>   add_stat_1 = "add_stat_1"
#> )
#> 
#> 
#> Column Name   Column Header      
#> ------------  -------------------
#> label         **Characteristic** 
#> stat_0        **N = 6,194**      
#> add_stat_1    add_stat_1
```

| **Characteristic** | **N = 6,194** | add_stat_1 |
|:-------------------|:-------------:|:----------:|
| awards2            |  4,400 (71%)  |            |

``` r

## change awards back to a char 
apiclus1 <- mutate(apiclus1, awards2 = as.character(awards2))
head(apiclus1$awards2)
#> [1] "TRUE"  "TRUE"  "TRUE"  "FALSE" "TRUE"  "TRUE"

## works 
tbl <-
  survey::svydesign(id = ~dnum, weights = ~pw, data = apiclus1, fpc = ~fpc) %>%
  tbl_svysummary(
    include = c(awards2),
    statistic = all_continuous() ~ "{mean}"
  )

tbl %>% 
  add_ci() %>% 
  as_kable()
```

| **Characteristic** | **N = 6,194** | **95% CI** |
|:-------------------|:-------------:|:----------:|
| awards2            |               |            |
| FALSE              |  1,794 (29%)  |  22%, 37%  |
| TRUE               |  4,400 (71%)  |  63%, 78%  |

<sup>Created on 2023-02-03 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: R/add_ci.R [hunk 1]
```diff
@@ -404,3 +478,202 @@ calculate_prop_ci <- function(x, n, statistic, method, conf.level, style_fun) {
...
+calculate_svyprop_ci <- function(variable, variable_levels, by, level, tbl,
+                                 method, statistic, conf.level, style_fun, df) {
+  if (is.null(by) || is.na(level))
+    design <- tbl$inputs$data
+  else
+    design <- subset(
+      tbl$inputs$data,
+      tbl$inputs$data$variables[[by]] == level
+    )
+
```
Comment (aspina7) [inline R/add_ci.R:657]: potential catch for dichotomous variables, to allow `calculate_svyprop_ci()` to run 
```r
if (!exists("tbl$meta_data$variable_levels") && tbl$meta_data$summary_type == "dichotomous") {
    variable_levels <- unlist(tbl$meta_data$dichotomous_value)
  }
```

**Comment by aspina7** [comment]
@ddsjoberg  @larmarange hi both just following up on my comment above - realised that the reason dichotomous variables were not working is that the tbl_svysummary() does not have a variable_levels column in the metadata. I chucked in an ifelse catch [here](https://github.com/ddsjoberg/gtsummary/pull/1402/files#r1096788540) and managed to run `add_ci()` on dichotomous variables. 
I managed to build locally - and the outputs below worked, but there were several tests and checks failing so didn't want to make a pr. Happy to open one if useful! 
Thanks again I really appreciate all the work you guys have put into the package! 

``` r
library(gtsummary)

## load data
data(api, package = "survey")

## run a svysummary with a "Yes"/"No" var 
tbl <-
  survey::svydesign(id = ~dnum, weights = ~pw, data = apiclus1, fpc = ~fpc) %>%
  tbl_svysummary(
    include = c(awards)
  )

## returns a single row table with only the "Yes" level
tbl %>% as_kable() 
```

| **Characteristic** | **N = 6,194** |
|:-------------------|:-------------:|
| awards             |  4,400 (71%)  |

``` r

## see that variable_levels missing 
tbl$meta_data %>% knitr::kable()
```

| variable | summary_type | dichotomous_value | var_label | stat_display | stat_label | sort         | df_stats                                                                                                                                                                                                                     |
|:---------|:-------------|:------------------|:----------|:-------------|:-----------|:-------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| awards   | dichotomous  | Yes               | awards    | {n} ({p}%)   | n (%)      | alphanumeric | awards , {n} ({p}%) , 130 , 183 , 0.710382513661202 , 4400.1095199585 , 0.710382513661202 , 0.0330182995537189, 6194.00032424927 , 0 , 183 , 0 , 0 , 0 , 6194.00032424927 , 6194.00032424927 , 1 , 183 , 1 , awards , stat_0 |

``` r

## able to add confidence intervals 
tbl %>% 
  add_ci() %>% 
  as_kable()
```

| **Characteristic** | **N = 6,194** | **95% CI** |
|:-------------------|:-------------:|:----------:|
| awards             |  4,400 (71%)  |  63%, 78%  |

<sup>Created on 2023-02-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by larmarange** [comment]
Hi, sorry I'm currently traveling but I will look at it as soon as I can

**Comment by larmarange** [comment]
@aspina7  Can you look at #1456 ?

**Comment by aspina7** [comment]
perfect thanks @larmarange !

**Pull request merged**